chia/rpc/data_layer_rpc_api.py

```diff
@@ -336,3 +554,230 @@ async def get_mirrors(self, request: Dict[str, Any]) -> EndpointResult:
+                proof = Proof(
+                    key=entry.key,
+                    value=entry.value,
+                    node_hash=proof_of_inclusion.node_hash,
+                    layers=tuple(
+                        Layer(
+                            other_hash_side=layer.other_hash_side,
+                            other_hash=layer.other_hash,
+                            combined_hash=layer.combined_hash,
+                        )
+                        for layer in proof_of_inclusion.layers
+                    ),
+                )
+                proofs.append(proof)
+            store_proof = StoreProofs(store_id=offer_store.store_id, proofs=tuple(proofs))
+            our_store_proofs[offer_store.store_id] = store_proof
+
+        offer_dict: Dict[Union[uint32, str], int] = {
+            **{offer_store.store_id.hex(): -1 for offer_store in request.maker},
+            **{offer_store.store_id.hex(): 1 for offer_store in request.taker},
+        }
+
+        solver: Dict[str, Any] = {
+            "0x"
+            + our_offer_store.store_id.hex(): {
+                "new_root": "0x" + our_store_proofs[our_offer_store.store_id].proofs[0].layers[-1].combined_hash.hex(),
+                "dependencies": [
+                    {
+                        "launcher_id": "0x" + their_offer_store.store_id.hex(),
+                        "values_to_prove": [
+                            "0x" + leaf_hash(key=entry.key, value=entry.value).hex()
+                            for entry in their_offer_store.inclusions
+                        ],
+                    }
+                    for their_offer_store in request.taker
+                ],
+            }
+            for our_offer_store in request.maker
+        }
+
+        fee = get_fee(self.service.config, {"fee": request.fee})
+
+        offer, trade_record = await self.service.wallet_rpc.create_offer_for_ids(
+            offer_dict=offer_dict,
+            solver=solver,
+            driver_dict={},
+            fee=fee,
+            validate_only=False,
+        )
+        if offer is None:
+            raise Exception("offer is None despite validate_only=False")
+
+        return MakeOfferResponse(
+            success=True,
+            offer=Offer(
+                offer_id=trade_record.trade_id,
+                offer=bytes(offer),
+                taker=request.taker,
+                maker=tuple(our_store_proofs.values()),
+            ),
+        )
+
+    @marshal()  # type: ignore[arg-type]
+    async def take_offer(self, request: TakeOfferRequest) -> TakeOfferResponse:
+        our_store_proofs: List[StoreProofs] = []
+        for offer_store in request.offer.taker:
+            changelist = await self.build_offer_changelist(
+                store_id=offer_store.store_id,
+                inclusions=offer_store.inclusions,
+            )
+
+            if len(changelist) > 0:
+                new_root_hash = await self.service.data_store.insert_batch(
```

Comment:
if something fails below, I think we'll still be stuck with the pending root indefinitely on the taker side (which would block further `insert_batch` operations). I think it would be nice to delete it if we had any failure below (for example with `data_store.clear_pending_roots`).

probably not critical tho, since the pending root will be deleted when maker finally produces something correct (since the singleton will change the value, pending roots that don't match the new value get deleted).